clients/src/main/java/org/apache/kafka/common/record/FileLogBuffer.java

```diff
@@ -0,0 +1,457 @@
+            Record shallowRecord = shallowEntry.record();
+            if (shallowRecord.timestamp() >= targetTimestamp) {
+                // We found a message
+                for (LogEntry deepLogEntry : shallowEntry) {
+                    long timestamp = deepLogEntry.record().timestamp();
+                    if (timestamp >= targetTimestamp)
+                        return new TimestampAndOffset(timestamp, deepLogEntry.offset());
+                }
+                throw new IllegalStateException(String.format("The message set (max timestamp = %s, max offset = %s" +
+                        " should contain target timestamp $targetTimestamp but it does not.", shallowRecord.timestamp(),
+                        shallowEntry.offset()));
+            }
+        }
+        return null;
+    }
+
+    /**
+     * Return the largest timestamp of the messages after a given position in this file message set.
+     * @param startingPosition The starting position.
+     * @return The largest timestamp of the messages after the given position.
+     */
+    public TimestampAndOffset largestTimestampAfter(int startingPosition) {
+        long maxTimestamp = Record.NO_TIMESTAMP;
+        long offsetOfMaxTimestamp = -1L;
+
+        Iterator<FileChannelLogEntry> shallowIterator = shallowEntriesFrom(startingPosition);
+        while (shallowIterator.hasNext()) {
+            LogEntry shallowEntry = shallowIterator.next();
+            long timestamp = shallowEntry.record().timestamp();
+            if (timestamp > maxTimestamp) {
+                maxTimestamp = timestamp;
+                offsetOfMaxTimestamp = shallowEntry.offset();
+            }
+        }
+        return new TimestampAndOffset(maxTimestamp, offsetOfMaxTimestamp);
+    }
+
+    @Override
+    public boolean hasMatchingShallowMagic(byte magic) {
+        Iterator<FileChannelLogEntry> iterator = shallowEntriesFrom(Integer.MAX_VALUE, start, false);
+        while (iterator.hasNext())
+            if (iterator.next().magic() != magic)
+                return false;
+        return true;
+    }
+
+    @Override
+    public Iterator<FileChannelLogEntry> shallowEntries() {
+        return shallowEntriesFrom(start);
+    }
+
+    public Iterator<FileChannelLogEntry> shallowEntries(int maxMessageSize) {
+        return shallowEntriesFrom(maxMessageSize, start, true);
+    }
+
+    private Iterator<FileChannelLogEntry> shallowEntriesFrom(long start) {
+        return shallowEntriesFrom(Integer.MAX_VALUE, start, true);
+    }
+
+    private Iterator<FileChannelLogEntry> shallowEntriesFrom(int maxMessageSize, long start, boolean eagerLoadRecords) {
+        final long end;
+        if (isSlice)
+            end = this.end;
+        else
+            end = this.sizeInBytes();
+        FileLogInputStream inputStream = new FileLogInputStream(channel, maxMessageSize, start, end, eagerLoadRecords);
+        return LogBufferIterator.shallowIterator(inputStream);
```

Comment:
Hmm.. there was actually a reason for this. The static `shallowIterator` returns the more specific LogEntry type, which would not be possible if `shallow` is passed as an argument.  Having the more specific type in shallow iteration lets you do some stuff to the shallow entries that are not possible with the deep entries (such as setting offsets or timestamps in-place).